pkg/dial/clustermesh_resolver.go

```diff
@@ -0,0 +1,128 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package dial
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/netip"
+
+	"github.com/cilium/hive/cell"
+	"github.com/cilium/hive/job"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+
+	k8sClient "github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+const (
+	clustermeshResolverConfigMapName string = "cilium-clustermesh-clusters"
+	clustermeshResolverConfigMapKey  string = "clusters.json"
+)
+
+// ClustermeshResolverCell provides a ClustermeshResolver instance to map DNS names
+// matching clustermesh hosts to IP addresses provided in a configmap.
+var ClustermeshResolverCell = cell.Module(
+	"clustermesh-resolver",
+	"ClustermeshResolver maps aliases for clustermeshes to their IP addresses",
+
+	cell.Provide(newClustermeshResolver),
+)
+
+var _ Resolver = (*ClustermeshResolver)(nil)
+
+type clustermeshConfig struct {
+	Domain   string `json:"domain"`
+	Clusters []struct {
+		Name    string       `json:"name"`
+		Address string       `json:"address"`
+		IPs     []netip.Addr `json:"ips"`
+	} `json:"clusters"`
+}
+
+// ClustermeshResolver maps aliases for clustermeshes to their IP addresses.
+type ClustermeshResolver struct {
+	mu      lock.RWMutex
+	records map[string]netip.Addr
+}
+
+func newClustermeshResolver(jg job.Group, client k8sClient.Clientset) *ClustermeshResolver {
```

Comment:
You have a component that is watching for changes to a single hard-coded ConfigMap name here. You don't need to do that :-). The kubelet will do it for you, writing any changes to a ConfigMap volume.

I would suggest writing an `/etc/hosts`-style parser that watches a directory for changes. Let the kubelet do the dirty work.

Reply:
I was suggesting that in the other comment - we also want to make sure that for a new cluster, fqdn <-> IP mapping is discovered at the same time as the cluster so we won't have race-conditions here.